pkg/apis/pipeline/v1beta1/matrix_types.go

```diff
@@ -56,37 +56,121 @@ type MatrixInclude struct {
 type Combination map[string]string
 
-// Combinations is a list of combination
+// Combinations is a Combination list
 type Combinations []Combination
 
+// FanOut generates Combinations, which is a list of mapped Combination by fanning out matrix
+// Parameters including Matrix Include Parameters
+func (m *Matrix) FanOut() Combinations {
+	var combinations, includeCombinations Combinations
+
+	// Generate include combinations from Matrix Include Parameters
+	includeCombinations = m.fanOutIncludeParams()
+
+	if m.hasInclude() && !m.hasParams() {
+		// If there are only Matrix Include Parameters return explicit combinations
+		return includeCombinations
+	}
+
+	// Generate combinations from Matrix Parameters
+	for _, parameter := range m.Params {
+		combinations = combinations.fanOutMatrixParams(parameter)
+	}
+
+	// Override the current combination with the include combination if any of the include parameters are
+	// found in combinations with matching parameter name and values
+	for _, paramCombination := range combinations {
+		for _, includeCombination := range includeCombinations {
+			if paramCombination.contains(includeCombination) {
+				for name, val := range includeCombination {
+					paramCombination[name] = val
+				}
+			}
+		}
+	}
+
+	// Generate a new combination for the missing include parameter value
+	for _, includeCombination := range includeCombinations {
+		if combinations.shouldAddNewCombination(includeCombination) {
+			combinations = append(combinations, includeCombination)
+		}
+	}
+
+	return combinations
+}
+
+// contains returns true if the include parameter name and value exists in combinations
+func (c Combination) contains(includeCombination Combination) bool {
+	for name, val := range includeCombination {
+		if _, exist := c[name]; exist {
+			if c[name] != val {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// shouldAddNewCombination returns true if the include parameter name exists but the value is
+// missing from combinations
+func (cs Combinations) shouldAddNewCombination(includeCombination map[string]string) bool {
+	if len(includeCombination) == 0 {
+		return false
+	}
+	for _, paramCombination := range cs {
+		for name, val := range includeCombination {
+			if _, exist := paramCombination[name]; exist {
+				if paramCombination[name] == val {
+					return false
+				}
+			}
+		}
+	}
+	return true
+}
+
 // ToParams transforms Combinations from a slice of map[string]string to a slice of Params
-// such that, the combinations can be directly consumed in creating taskRun/run object
+// such that, these combinations can be directly consumed in creating taskRun/run object
 func (cs Combinations) ToParams() []Params {
 	listOfParams := make([]Params, len(cs))
 	for i := range cs {
 		var params Params
 		combination := cs[i]
-		order, sortedCombination := combination.sortCombination()
+		order, _ := combination.sortCombination()
```

Comment:
if the return values aren't used, can we update this function to only return the values needed?

Reply:
This function is being called in other places where it uses the sortedCombination.

```// distribute generates a new Combination of Parameters by adding a new Parameter to an existing list of Combinations.
func (cs Combinations) distribute(param Param) Combinations {
	var expandedCombinations Combinations
	for _, value := range param.Value.ArrayVal {
		for _, combination := range cs {
			newCombination := make(Combination)
			maps.Copy(newCombination, combination)
			newCombination[param.Name] = value
			_, orderedCombination := newCombination.sortCombination()
			expandedCombinations = append(expandedCombinations, orderedCombination)
		}
	}
	return expandedCombinations
}